src/toil/lib/aws/__init__.py

```diff
@@ -11,22 +11,258 @@
+    stop and return that answer. Otherwise, return None.
+    """
+
+    for callback in functions:
+        # Call all the passed functions in order
+
+        # We need to work out what arguments are accepted and out them all in here
+        call_args: Dict[str, Any] = {}
+        sig = inspect.signature(callback)
+        for param_name, param in sig.parameters:
+            # For each accepted parameter
+            if param.kind in [inspect.Parameter.POSITIONAL_ONLY, inspect.Parameter.VAR_POSITIONAL]:
+                # These can't be kwargs, so ignore them.
+                pass
+            elif param.kind == inspect.Parameter.VAR_KEYWORD:
+                # Function can take any kwarg, so dump them all in.
+                call_args.update(kwargs)
+                break
+            elif param_name in kwargs:
+                # This is just one parameter that can be a kwarg, and we have a value for it
+                call_args[param_name] = kwargs[param_name]
+
+        # Now pass the subset of kwargs along to the function
+        result = callback(**call_args)
+        if result is not None:
+            # And return the first non-None result
+            return result
+    # Otherwise return None
+    return None
+
+def get_aws_zone_from_environment() -> Optional[str]:
+    """
+    Get the AWS zone from TOIL_AWS_ZONE if set.
+    """
+    return os.environ.get('TOIL_AWS_ZONE', None)
+
+def get_aws_zone_from_metadata() -> Optional[str]:
+    """
+    Get the AWS zone from instance metadata, if on EC2 and the boto module is
+    available.
+    """
+    if running_on_ec2():
+        try:
+            import boto
+            from boto.utils import get_instance_metadata
+            return get_instance_metadata()['placement']['availability-zone']
+        except (KeyError, ImportError):
+            pass
+    return None
+
+def get_aws_zone_from_boto() -> Optional[str]:
+    """
+    Get the AWS zone from the Boto config file, if it is configured and the
+    boto module is avbailable.
+    """
+    try:
+        import boto
+        zone = boto.config.get('Boto', 'ec2_region_name')
+        if zone is not None:
+            zone += 'a'  # derive an availability zone in the region
+        return zone
+    except ImportError:
+        pass
+    return None
+
+
+def get_current_aws_zone() -> Optional[str]:
+    """
+    Get the currently configured or occupied AWS zone to use.
+
+    Reports the TOIL_AWS_ZONE environment variable if set.
+
+    Otherwise, if we have boto and are running on EC2, reports the zone we are
+    running in.
+
+    Finally, if we have boto2, and a default region is configured in Boto 2,
+    chooses a zone in that region.
+
+    Returns None if no method can produce a zone to use.
+    """
+    return try_providers([get_aws_zone_from_environment,
+                          get_aws_zone_from_metadata,
+                          get_aws_zone_from_boto])
```

Comment:
Hmmm... why this over something like:

```
return get_aws_zone_from_environment() or get_aws_zone_from_metadata() or get_aws_zone_from_boto() or None
```

Reply:
Yeah I think I can actually just use `or` in both places I am using this.